delivery_postlogistics/postlogistics/web_service.py

```diff
@@ -0,0 +1,492 @@
+        ]
+        if is_phone_required:
+            phone = picking.delivery_phone or partner.phone
+            if phone:
+                recipient["phone"] = phone
+
+            mobile = picking.delivery_mobile or partner.mobile
+            if mobile:
+                recipient["mobile"] = mobile
+
+        return recipient
+
+    def _prepare_customer(self, picking):
+        """ Create a ns0:Customer as a dict from picking
+
+        This is the Postlogistic Customer, thus the sender
+
+        :param picking: picking browse record
+        :return a dict containing data for ns0:Customer
+
+        """
+        company = picking.company_id
+        partner = company.partner_id
+
+        customer = {
+            "name1": partner.name,
+            "street": partner.street,
+            "zip": partner.zip,
+            "city": partner.city,
+            "country": partner.country_id.code,
+            "domicilePostOffice": picking.carrier_id.postlogistics_office,
+        }
+        logo = picking.carrier_id.postlogistics_logo
+        if logo:
+            logo_image = Image.open(BytesIO(base64.b64decode(logo)))
+            logo_format = logo_image.format
+            customer["logo"] = logo.decode()
+            customer["logoFormat"] = logo_format
+        return customer
+
+    def _get_single_option(self, picking, option):
+        option = [
+            opt.code
+            for opt in picking.postlogistics_option_ids
+            if opt.postlogistics_type == option
+        ]
+        assert len(option) <= 1
+        return option and option[0]
+
+    def _get_label_layout(self, picking):
+        label_layout = self._get_single_option(picking, "label_layout")
+        if not label_layout:
+            label_layout = picking.carrier_id.postlogistics_label_layout.code
+        return label_layout
+
+    def _get_output_format(self, picking):
+        output_format = self._get_single_option(picking, "output_format")
+        if not output_format:
+            output_format = picking.carrier_id.postlogistics_output_format.code
+        return output_format
+
+    def _get_image_resolution(self, picking):
+        resolution = self._get_single_option(picking, "resolution")
+        if not resolution:
+            resolution = picking.carrier_id.postlogistics_resolution.code
+        return resolution
+
+    def _get_license(self, picking):
+        """ Get the license
+
+        Take it from carrier and if not defined get the first license.
+
+        :return: license number
+        """
+        franking_license = picking.carrier_id.postlogistics_license_id
+        # if not franking_license:
+        #     company_licenses = picking.company_id.postlogistics_license_ids
+        #     if not company_licenses:
+        #         return None
+        #     franking_license = company_licenses[0]
+        return franking_license.number
+
+    def _prepare_attributes(self, picking, pack_num=None, pack_total=None):
```

Comment:
Add pack in the attributes
self, picking, pack=None, pack_num=0, pack_total=0

Regarding PRZL (service), we need to be able to choose it when creating the package.
For this, there is the concept of product.packaging that is defined in the delivery module. When a package is created a product.packaging is assigned to it. So, we need to load a list of product.packaging for postlogistics with all the possible service values. Then we need to be able to define a default postlogistics product packaging (i.e. default service) on the carrier.
If there is no package, the value PRZL is the carrier default service. Else for each package, is it is the service defined on the assigned product packaging.
This also means that the service will not be defined by an option code anymore.

Reply:
I dont really understand this. Do you mean that we can use the `product.packaging` generated when creating a package for defining the services przl ?
What is the default carrier service ? For the assigning product.packaging, which values should be used ? `Package Code` I think

Reply:
All the product.packaging possibilities must be pre-loaded with a data file and not created during processing. This defines the service that is used.
At SO, the post carrier is chosen and on that carrier, for instance, it defaults to PRI service (default product.packaging on the carrier). Then at delivery when you create the package, you can still select a product.packaging (for instance SP service due to over sized package). So you assign that product.packaging to the package (if you do not assign a specific one, it defaults to the default value of the carrier). Based on the product.packaging, you know what is the service to declare.
DHL enterprise module is working similarly.

Reply:
Understand, will update the source code for this soon. Thanks

Reply:
@jbaudoux as far as I check, each move of the picking will be assigned to a package (stock.quant.package) and a product packaging type (product.packaging). Therefore if we want to define multiple services for each picking, we may need to define multiple services in product packaging level. Correct me if I am wrong.
Btw, when I do the test with PRI (only one service: PRI), I got this error message.
```
Swisspost API returns errors:
Error code: E2030
Error message: Basic service PostPac Priority does not belong to an international service group.
```
Do you have any idea?

Reply:
@phuctrphuctranerp Indeed, each package (stock.quant.package) has a product packaging that has a defined service. So each package has indirectly a service associated.

Regarding the error, maybe in your test you didn't used a from and to address in Switzerland?

Reply:
I have adjusted the address to have the package delivered internally in Switzerland. However, then I get the error 500 - Internal Server Error. Not sure why. Here is the json data I posted
URL: https://wedecint.post.ch/api/barcode/v1/generateAddressLabel
Json Data
```
{
    "language": "DE",
    "frankingLicense": "60013217",
    "customer": {
        "name1": "A Swiss Company",
        "street": "Rue de Lausanne 1",
        "zip": "1030",
        "city": "Bussigny",
        "country": "CH"
    },
    "labelDefinition": {
        "labelLayout": "A7",
        "printAddresses": "RECIPIENT_AND_CUSTOMER",
        "imageFileType": "PNG",
        "imageResolution": "200"
    },
    "item": {
        "itemID": "WHOUT00005+PACK0000005",
        "recipient": {
            "name1": "Wood Corner",
            "street": "EPFL Innovation Park, B\u00e2t A",
            "zip": "1015",
            "city": "Lausanne",
            "email": "wood.corner26@example.com",
            "country": "CH"
        },
        "attributes": {
            "przl": [
                "PRI"
            ],
            "weight": 0.16
        }
    }
}
```

I will recheck it again, it could be an issue of testing server, or do you have any idea? @jbaudoux ?

Reply:
works well now by changing the end point url to https://wedec.post.ch/

Reply:
This is still not good when you have multiple packs. From what I understand and as I wrote initially, you need to know for which pack you are generating the attributes: self, picking, **pack=None**, pack_num=0, pack_total=0

In case of multiple packs, picking._get_picking_postlogistic_packaging() is not returning the product packaging of the processed pack.


Then regarding combined DLC like for instance "SP, PRI" for "Bulky goods Priority", as there is only one shipper_package_code on the product packaging, we could define it as "SP, PRI" but then you need here to split it to declare the service as ["SP","PRI"]

Reply:
Hi, sorry I missed this comment. Thanks for your recommendation, will be update soon.

Reply:
https://github.com/OCA/delivery-carrier/pull/323/commits/f04b053b0aeca3c51756f5ad9129349ce26f4555